File: src/main/java/com/abysalto/middleware/config/CacheKeys.java
package com.abysalto.middleware.config;

import java.math.BigDecimal;
import java.util.Locale;

/**
 * Builds normalized cache keys for the search/filter caches so parameter sets that are semantically
 * identical map to the same entry: case/whitespace differences in text, equivalent numeric scales
 * (e.g. {@code 10} vs {@code 10.00}) and absent price bounds all collapse to one key.
 *
 * <p>{@link #normalizeText(String)} is the single source of truth for free-text normalization: the
 * {@code @Cacheable} SpEL keys in {@code ProductService} and the service body both run the query and
 * category through it before use, so the cache key and the actual upstream call can never diverge
 * (two inputs share a key only when they also produce the same upstream call).
 */
public final class CacheKeys {

    private CacheKeys() {
    }

    /** Normalizes a free-text param (query/category): trimmed and lower-cased; null/blank become "". */
    public static String normalizeText(String value) {
        return value == null ? "" : value.strip().toLowerCase(Locale.ROOT);
    }

    /** Key for name search: normalized query plus pagination. */
    public static String search(String query, int page, int size) {
        return normalizeText(query) + "|" + page + "|" + size;
    }

    /** Key for filtering: normalized category and price bounds plus pagination. */
    public static String filter(String category, BigDecimal minPrice, BigDecimal maxPrice, int page, int size) {
        return normalizeText(category) + "|" + price(minPrice) + "|" + price(maxPrice) + "|" + page + "|" + size;
    }

    private static String price(BigDecimal value) {
        return value == null ? "*" : value.stripTrailingZeros().toPlainString();
    }
}

File: dotnet/src/Middleware.Core/Common/CacheKeys.cs
using System.Globalization;

namespace Middleware.Core.Common;

/// <summary>
/// Builds normalized cache keys for the search/filter caches so parameter sets that are semantically
/// identical map to the same entry: case/whitespace differences in text, equivalent numeric scales
/// (e.g. <c>10</c> vs <c>10.00</c>) and absent price bounds all collapse to one key.
/// <para><see cref="NormalizeText"/> is the single source of truth for free-text normalization: the
/// cache key and the actual upstream call both run the query and category through it, so they can
/// never diverge (two inputs share a key only when they also produce the same upstream call).</para>
/// </summary>
public static class CacheKeys
{
    /// <summary>Normalizes a free-text param (query/category): trimmed and lower-cased; null/blank become "".</summary>
    public static string NormalizeText(string? value) =>
        value is null ? string.Empty : value.Trim().ToLowerInvariant();

    /// <summary>Key for name search: normalized query plus pagination.</summary>
    public static string Search(string? query, int page, int size) =>
        "search|" + NormalizeText(query) + "|" + page + "|" + size;

    /// <summary>Key for the no-price filter path (category listing or full list): normalized category plus pagination.</summary>
    public static string FilterPage(string? category, int page, int size) =>
        "page|" + NormalizeText(category) + "|" + page + "|" + size;

    /// <summary>
    /// Key for the price-filter candidate set: normalized category plus price bounds only —
    /// deliberately independent of pagination, so every page of the same filter reuses one cached
    /// upstream fetch. The <c>page|</c>/<c>cand|</c> prefixes keep the two filter-cache key spaces
    /// from ever colliding.
    /// </summary>
    public static string FilterCandidates(string? category, decimal? minPrice, decimal? maxPrice) =>
        "cand|" + NormalizeText(category) + "|" + Price(minPrice) + "|" + Price(maxPrice);

    // Mirrors BigDecimal.stripTrailingZeros().toPlainString(): fixed-point, no trailing zeros, no
    // exponent. decimal preserves scale (10.00m renders as "10.00"), so strip trailing zeros here.
    private static string Price(decimal? value)
    {
        if (value is null)
        {
            return "*";
        }
        string s = value.Value.ToString(CultureInfo.InvariantCulture);
        if (s.Contains('.'))
        {
            s = s.TrimEnd('0').TrimEnd('.');
        }
        return s;
    }
}
